Sources/NIOSSL/SSLContext.swift

```diff
@@ -779,6 +791,90 @@ extension NIOSSLContext {
+
+                // The user may or may not retain a reference to the NIOSSLContext
+                // that they return here, if they provide one. The NIOSSLContext finalizer
+                // explicitly decrements the reference count on the underlying SSL_CTX.
+                //
+                // However, SSL_set_SSL_CTX explicitly increases the reference count and
+                // will decrement the count when the SSL* object is destroyed, thus finally
+                // freeing the raw SSL_CTX object, even if the swift reference is orphaned.
+                //
+                // As such, no memory should be leaked or freed too early in the event that
+                // the user fails to retain their created and return NIOSSLContext. In the
+                // event that they do retain it, the reference count will remain at at-least
+                // 1 and thus they can re-use the same context for multiple connections.
+                let nativeContextValue = withExtendedLifetime(userChosenContext) {
+                    return CNIOBoringSSL_SSL_set_SSL_CTX(ssl, userChosenContext.sslContext)
+                }
+
+                // CNIOBoringSSL_SSL_set_SSL_CTX should return a pointer to the very same context it was
+                // told to assign on success. On error, this will not happen, so we check for equality
+                // here and handle any such errors by signaling the connection to fail.
+                if nativeContextValue != userChosenContext.sslContext {
+                    return SSL_TLSEXT_ERR_NOACK
+                }
+
+                return SSL_TLSEXT_ERR_OK
+            }
+
+            // Make sure we get the server name extracted.
+            guard let cServerName = CNIOBoringSSL_SSL_get_servername(ssl, CNIOBoringSSL.TLSEXT_NAMETYPE_host_name) else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            let serverName = String(cString: cServerName)
+
+            // Can do nothing about empty indicators.
+            guard !serverName.isEmpty else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            // Construct extension values
+            let values = NIOSSLClientExtensionValues(serverName: serverName)
+
+            // Issue call to get a new ssl context
+            let futureSSLContext: EventLoopFuture<NIOSSLContext>
+            do {
+                futureSSLContext = try parentSwiftContext.sslContextCallback!(values, parentSwiftContext)
+            } catch {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            // Save the future as pending on the parent context
+            parentSwiftContext.futureSSLContext = futureSSLContext
+
+            // Resume the handshake when the future completes
+            futureSSLContext.whenComplete { _ in
+                CNIOBoringSSL_SSL_do_handshake(ssl)
+            }
+
+            // We must return a negative value to suspend the handshake
+            return -1
```

Comment:
You'll also need to add `wantX509Lookup` to the list of reason codes that cause us to assume a handshake is incomplete in `SSLConnection.doHandshake`.